src/emit/solana/mod.rs

```diff
@@ -1127,4 +1128,252 @@ impl SolanaTarget {
+
+        binary.builder.build_store(
+            binary
+                .builder
+                .build_struct_gep(instruction_ty, instruction, 4, "data_len")
+                .unwrap(),
+            binary.builder.build_int_z_extend(
+                payload_len,
+                binary.context.i64_type(),
+                "payload_len",
+            ),
+        );
+
+        let parameters = self.sol_parameters(binary);
+
+        let account_infos = binary
+            .builder
+            .build_struct_gep(
+                binary
+                    .module
+                    .get_struct_type("struct.SolParameters")
+                    .unwrap(),
+                parameters,
+                0,
+                "ka",
+            )
+            .unwrap();
+
+        let account_infos_len = binary.builder.build_int_truncate(
+            binary
+                .builder
+                .build_load(
+                    binary.context.i64_type(),
+                    binary
+                        .builder
+                        .build_struct_gep(
+                            binary
+                                .module
+                                .get_struct_type("struct.SolParameters")
+                                .unwrap(),
+                            parameters,
+                            1,
+                            "ka_num",
+                        )
+                        .unwrap(),
+                    "ka_num",
+                )
+                .into_int_value(),
+            binary.context.i32_type(),
+            "ka_num",
+        );
+
+        let external_call = binary.module.get_function("sol_invoke_signed_c").unwrap();
+
+        let (signer_seeds, signer_seeds_len) = if let Some((seeds, len)) = contract_args.seeds {
+            (
+                seeds,
+                binary.builder.build_int_cast(
+                    len,
+                    external_call.get_type().get_param_types()[4].into_int_type(),
+                    "len",
+                ),
+            )
+        } else {
+            (
+                external_call.get_type().get_param_types()[3]
+                    .const_zero()
+                    .into_pointer_value(),
+                external_call.get_type().get_param_types()[4]
+                    .const_zero()
+                    .into_int_value(),
+            )
+        };
+
+        let ret = binary
+            .builder
+            .build_call(
+                external_call,
+                &[
+                    instruction.into(),
+                    account_infos.into(),
+                    account_infos_len.into(),
+                    signer_seeds.into(),
+                    signer_seeds_len.into(),
+                ],
+                "",
+            )
+            .try_as_basic_value()
+            .left()
+            .unwrap()
+            .into_int_value();
+
+        self.return_error(function, binary, ret, error_message, ns, loc);
```

Comment:
I think `sol_invoke_signed_c` cannot fail on Solana. It either returns success or the  transaction is aborted. So, this code is never executed (the error path).

Reply:
> I think

Are you not sure? We should be sure that is does not fail in order to remove this code path.

Reply:
Please look at the solana source code.

Reply:
If I am checking this correctly, the function can return something other than zero if the I pass invalid data to it. In this case, we can have a return without aborting.

https://github.com/solana-labs/solana/blob/a27ff1c65efe7858164255362a938cced6d4f055/programs/bpf_loader/src/syscalls/cpi.rs#L925

Reply:
I'll create a test to eliminate all my doubts.

Reply:
> If I am checking this correctly, the function can return something other than zero if the I pass invalid data to it. In this case, we can have a return without aborting.
> 
> https://github.com/solana-labs/solana/blob/a27ff1c65efe7858164255362a938cced6d4f055/programs/bpf_loader/src/syscalls/cpi.rs#L925

Where exactly do you see this path?

Reply:
`cpi_common` is registered as the sys call for SolInvokeSignedC: https://github.com/solana-labs/solana/blob/a27ff1c65efe7858164255362a938cced6d4f055/programs/bpf_loader/src/syscalls/cpi.rs#L492-L514

Reply:
So where exactly do you see this error path where non-zero is returned?

Reply:
I was sure I have seen the error message `failed to invoke sol_invoke_signed_c` I output there in case the function does not return zero while I was getting the PR ready, but I could not recreate any case that fails today. I'll remove the branch.